x/post/manager/manager.go

```diff
@@ -0,0 +1,296 @@
+	"github.com/lino-network/lino/utils"
+	acc "github.com/lino-network/lino/x/account"
+	dev "github.com/lino-network/lino/x/developer"
+	global "github.com/lino-network/lino/x/global"
+	"github.com/lino-network/lino/x/post/model"
+	"github.com/lino-network/lino/x/post/types"
+	price "github.com/lino-network/lino/x/price"
+	rep "github.com/lino-network/lino/x/reputation"
+)
+
+type PostManager struct {
+	postStorage model.PostStorage
+
+	// deps
+	am    acc.AccountKeeper
+	gm    global.GlobalKeeper
+	dev   dev.DeveloperKeeper
+	rep   rep.ReputationKeeper
+	price price.PriceKeeper
+}
+
+// NewPostManager - create a new post manager
+func NewPostManager(key sdk.StoreKey, am acc.AccountKeeper, gm global.GlobalKeeper, dev dev.DeveloperKeeper, rep rep.ReputationKeeper, price price.PriceKeeper) PostManager {
+	return PostManager{
+		postStorage: model.NewPostStorage(key),
+		am:          am,
+		gm:          gm,
+		dev:         dev,
+		rep:         rep,
+		price:       price,
+	}
+}
+
+// DoesPostExist - check if post exist
+func (pm PostManager) DoesPostExist(ctx sdk.Context, permlink linotypes.Permlink) bool {
+	return pm.postStorage.HasPost(ctx, permlink)
+}
+
+// GetPost - return post.
+func (pm PostManager) GetPost(ctx sdk.Context, permlink linotypes.Permlink) (model.Post, sdk.Error) {
+	post, err := pm.postStorage.GetPost(ctx, permlink)
+	if err != nil {
+		return model.Post{}, err
+	}
+	return *post, nil
+}
+
+// CreatePost validate and handles CreatePostMsg
+// stateful validation;
+// 1. both author and post id exists.
+// 2. if createdBy is not author, then it must be an app.
+// 3. post does not exists.
+func (pm PostManager) CreatePost(ctx sdk.Context, author linotypes.AccountKey, postID string, createdBy linotypes.AccountKey, content string, title string) sdk.Error {
+	if !pm.am.DoesAccountExist(ctx, author) {
+		return types.ErrAccountNotFound(author)
+	}
+	if !pm.am.DoesAccountExist(ctx, createdBy) {
+		return types.ErrAccountNotFound(createdBy)
+	}
+	// if created by app, then app must exist.
+	if author != createdBy && !pm.dev.DoesDeveloperExist(ctx, createdBy) {
+		return types.ErrDeveloperNotFound(createdBy)
+	}
+	permlink := linotypes.GetPermlink(author, postID)
+	if pm.DoesPostExist(ctx, permlink) {
+		return types.ErrPostAlreadyExist(permlink)
+	}
+
+	createdAt := ctx.BlockHeader().Time.Unix()
+	postInfo := &model.Post{
+		PostID:    postID,
+		Title:     title,
+		Content:   content,
+		Author:    author,
+		CreatedBy: createdBy,
+		CreatedAt: createdAt,
+		UpdatedAt: createdAt,
+	}
+	pm.postStorage.SetPost(ctx, postInfo)
+	return nil
+}
+
+// UpdatePost - update post title, content and links.
+// stateful validation:
+// 1. author exist.
+// 2. post exist.
+func (pm PostManager) UpdatePost(ctx sdk.Context, author linotypes.AccountKey, postID, title, content string) sdk.Error {
+	if !pm.am.DoesAccountExist(ctx, author) {
+		return types.ErrAccountNotFound(author)
+	}
+	permlink := linotypes.GetPermlink(author, postID)
```

Comment:
If author is primary key here, update post shouldn't be able to change author, otherwise based on the code change author will never success.

Reply:
i don't think author is changed.